### PR TITLE
Exception suppression

### PR DESCRIPTION
**Pull request by ralight**

### CODE REVIEW

File: src/paho/mqtt/client.py [hunk 1]
```diff
@@ -2075,6 +2072,8 @@ def _call_socket_open(self):
                     except Exception as err:
                         self._easy_log(
                             MQTT_LOG_ERR, 'Caught exception in on_socket_open: %s', err)
+                        if self.suppress_exceptions == False:
```
Comment (swails) [inline src/paho/mqtt/client.py:2075]: `if not self.suppress_exceptions`

File: src/paho/mqtt/client.py [hunk 2]
```diff
@@ -2107,6 +2106,8 @@ def _call_socket_close(self, sock):
                     except Exception as err:
                         self._easy_log(
                             MQTT_LOG_ERR, 'Caught exception in on_socket_close: %s', err)
+                        if self.suppress_exceptions == False:
```
Comment (swails) [inline src/paho/mqtt/client.py:2109]: `if not self.suppress_exceptions`

File: src/paho/mqtt/client.py [hunk 3]
```diff
@@ -2142,6 +2143,8 @@ def _call_socket_register_write(self):
                 except Exception as err:
                     self._easy_log(
                         MQTT_LOG_ERR, 'Caught exception in on_socket_register_write: %s', err)
+                    if self.suppress_exceptions == False:
```
Comment (swails) [inline src/paho/mqtt/client.py:2146]: `if not self.suppress_exceptions`

File: src/paho/mqtt/client.py [hunk 4]
```diff
@@ -2178,6 +2181,8 @@ def _call_socket_unregister_write(self, sock=None):
                 except Exception as err:
                     self._easy_log(
                         MQTT_LOG_ERR, 'Caught exception in on_socket_unregister_write: %s', err)
+                    if self.suppress_exceptions == False:
```
Comment (swails) [inline src/paho/mqtt/client.py:2184]: `if not self.suppress_exceptions`

File: src/paho/mqtt/client.py [hunk 5]
```diff
@@ -2365,6 +2360,8 @@ def _packet_write(self):
                                     except Exception as err:
                                         self._easy_log(
                                             MQTT_LOG_ERR, 'Caught exception in on_publish: %s', err)
+                                        if self.suppress_exceptions == False:
```
Comment (swails) [inline src/paho/mqtt/client.py:2363]: `if not self.suppress_exceptions`

File: src/paho/mqtt/client.py [hunk 6]
```diff
@@ -3048,6 +3031,8 @@ def _handle_connack(self):
                     except Exception as err:
                         self._easy_log(
                             MQTT_LOG_ERR, 'Caught exception in on_connect: %s', err)
+                        if self.suppress_exceptions == False:
```
Comment (swails) [inline src/paho/mqtt/client.py:3034]: `if not self.suppress_exceptions`

File: src/paho/mqtt/client.py [hunk 7]
```diff
@@ -3167,6 +3152,8 @@ def _handle_suback(self):
                     except Exception as err:
                         self._easy_log(
                             MQTT_LOG_ERR, 'Caught exception in on_subscribe: %s', err)
+                        if self.suppress_exceptions == False:
```
Comment (swails) [inline src/paho/mqtt/client.py:3155]: `if not self.suppress_exceptions`

File: src/paho/mqtt/client.py [hunk 8]
```diff
@@ -3354,8 +3341,26 @@ def _handle_unsuback(self):
                     except Exception as err:
                         self._easy_log(
                             MQTT_LOG_ERR, 'Caught exception in on_unsubscribe: %s', err)
+                        if self.suppress_exceptions == False:
```
Comment (swails) [inline src/paho/mqtt/client.py:3344]: `if not self.suppress_exceptions`

File: src/paho/mqtt/client.py [hunk 9]
```diff
@@ -3354,8 +3341,26 @@ def _handle_unsuback(self):
...
+                    try:
+                        if properties:
+                            self.on_disconnect(
+                                self, self._userdata, rc, properties)
+                        else:
+                            self.on_disconnect(self, self._userdata, rc)
+                    except Exception as err:
+                        self._easy_log(
+                            MQTT_LOG_ERR, 'Caught exception in on_disconnect: %s', err)
+                        if self.suppress_exceptions == False:
```
Comment (swails) [inline src/paho/mqtt/client.py:3361]: `if not self.suppress_exceptions`

File: src/paho/mqtt/client.py [hunk 10]
```diff
@@ -3365,6 +3370,8 @@ def _do_on_publish(self, mid):
                     except Exception as err:
                         self._easy_log(
                             MQTT_LOG_ERR, 'Caught exception in on_publish: %s', err)
+                        if self.suppress_exceptions == False:
```
Comment (swails) [inline src/paho/mqtt/client.py:3373]: `if not self.suppress_exceptions`

File: src/paho/mqtt/client.py [hunk 11]
```diff
@@ -3424,6 +3431,8 @@ def _handle_on_message(self, message):
                                 callback.__name__,
                                 err
                             )
+                            if self.suppress_exceptions == False:
```
Comment (swails) [inline src/paho/mqtt/client.py:3434]: `if not self.suppress_exceptions`

File: src/paho/mqtt/client.py [hunk 12]
```diff
@@ -3433,6 +3442,8 @@ def _handle_on_message(self, message):
                     except Exception as err:
                         self._easy_log(
                             MQTT_LOG_ERR, 'Caught exception in on_message: %s', err)
+                        if self.suppress_exceptions == False:
```
Comment (swails) [inline src/paho/mqtt/client.py:3445]: `if not self.suppress_exceptions`

File: src/paho/mqtt/client.py (hunk 1 above)
Comment (ralight) [reply src/paho/mqtt/client.py]: All fixed, thanks.

**Comment by Cediddi** [comment]
Is there anything blocking this PR? Can I help? As this stands, there's no workaround to propagate any exception.

**Comment by Simanas** [comment]
Up! We need this as well! Please merge it in as soon as possible it's a critical fix!

**Comment by Simanas** [comment]
Thank you guys! This was really a crucial fix. Original implementation was extremely non Pythonic. Really against the core Zen of Python, where it says: "Errors should never pass silently. Unless explicitly silenced."
Thank's again for bringing this in!

**Pull request merged**